File: leetcode/2370-longest-ideal-subsequence.js
// https://leetcode.com/problems/longest-ideal-subsequence/

/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var longestIdealString = function(s, k) {
  // Time Complexity  : O(n*26) => O(n)
  // Space Complexity : O(26) => O(1)

  const n = s.length;
  
  const hashmap = {};
  hashmap[s.charCodeAt(n-1)] = 1;

  for (let i = n - 2; i >= 0; i -= 1) {
    let max = 1;
    for (let val = Math.max(97, s.charCodeAt(i) - k); val <= Math.min(122, s.charCodeAt(i) + k); val += 1) {
      if (Object.hasOwn(hashmap, val)) max = Math.max(max, (1 + hashmap[val]));
    }
    hashmap[s.charCodeAt(i)] = max;
  }

  return Math.max(...Object.values(hashmap));

};

File: leetcode/2370-longest-ideal-subsequence.py
# https://leetcode.com/problems/longest-ideal-subsequence/

class Solution:
    def longestIdealString(self, s: str, k: int) -> int:
        n = len(s)
        dp = [1] * n
        hashmap = {}
        ascii = []
        for c in s:
            ascii.append(int(ord(c)) - 96)
                
        hashmap[ascii[n-1]] = 1

        for i in reversed(range(n - 1)):
            for j in range(max(1, ascii[i] - k), 1 + min(26, ascii[i] + k), 1):
                if j in hashmap:
                    dp[i] = max(dp[i], (1 + hashmap[j]))
            
            hashmap[ascii[i]] = max(dp[i], (hashmap[ascii[i]] if (ascii[i] in hashmap ) else 1))

        return max(dp)
